content/2020-05-22-book-review-writing-an-interpreter-in-go/index.md

```diff
@@ -0,0 +1,581 @@
+    - [environment](#environment)
+    - [first-class citizens](#first-class-citizens)
+  - [Chapter 4 - Extending the Interpreter](#chapter-4---extending-the-interpreter)
+    - [built-in functions](#built-in-functions)
+    - [validate hash keys in evaluation stage](#validate-hash-keys-in-evaluation-stage)
+  - [Chapter 5 - Macro System](#chapter-5---macro-system)
+- [Overall Thoughts](#overall-thoughts)
+- [Recommend the Book?](#recommend-the-book)
+- [Purchase the Book](#purchase-the-book)
+
+## Notes
+
+This is a collection of my notes from each chapter in the book, along with my overall thoughts and recommendations.
+
+### Introduction
+
+In the beginning of the book, Thorsten explains why he wrote this book and who it's for. One thing that stuck out to me was this,
+
+> So I wrote this book, for you and me. This is the book I wish I had. This is a book for people who love to look under the hood. For people that love to learn by understanding how something really works.
+>
+> -Ball 7
+
+The part that resonated with me was the last sentence in hat quote -- "people that love to learn by understanding how something really works." That made me glad I picked this book up.
+
+> "Tree-walking interpreter"
+
+I highlighted this because I know I'll forgot what it means if I don't write it down. Summarized, it means an interpreter that parses the source code, builds an abstract syntax tree (AST) and then evaluates it. The "tree" comes from AST. The "walking" comes from walking down the tree.
+
+Lexer vs. parser?
+
+I felt confused in the beginning what the difference once. In my own words, the lexer takes input like characters "1;" and goes, "Ah! Yes. That input is equal to these tokens {token.INTEGER, '1', token.SEMICOLON, ';'}" The parser says, "Ah! Thank you lexer for the tokens, I will turn those into an AST."
+
+> Every interpreter is built to interpret a specific programming language. That's how you "implement" a programming language. Without a compiler or an interpreter, a programming language is nothing more than an idea or a specification.
+>
+> -Ball 8
+
+This was an aha moment for me. The compiler and the interpreter are key. I had no idea. I had been writing JavaScript for so long and never really looked under the hood to understand how the interpreter worked. I'm also learning Rust, which is my first compiled language, so hearing this helped my understanding there as well.
+
+#### hashes vs hashmaps vs dictionaries vs objects
+
+There is probably more to it than I know, but I find that this data structure has a different name depending on which language in which you're writing it:
+
+```javascript
+let previte = { name: 'Joe', age: 26 }
+```
+
+In the book (which uses Go), the author refers to it as a hash. In JavaScript, I would call this an object. In Python, one might call it a dictionary. Why can't we all decide on one name and use it across all languages? If someone knows the answer to this, please fill me in!
```

Comment:
Imagine me waving my hand here, slightly moving my head and saying "Hmm, well, it's not that easy..."

What Python calls dictionaries is what Go calls maps and what's usually referred to as hash table (hash for short, or hash map): https://en.wikipedia.org/wiki/Hash_table

JavaScript objects are their own datatype (remember, they also have methods and attributes, which hash tables do _not_ have). Under the hood, a JavaScript object can be implemented with hash tables, though. But that's just an implementation detail when we're talking about whether JavaScript objects are hash tables: they're not.

But what gets everyone confused is that their _literal_ syntax looks really similar (`map[string]string{"foobar": "barfoo"}` in Go and `{"foobar": "barfoo"}` in JavaScript) and that hash tables in other languages often encode to objects in JSON.

Reply:
🤯 OH! That makes a lot more sense - I'm going to include this in the post and quote you! (as long as that's cool!)